packages/flutter/lib/src/rendering/object.dart

```diff
@@ -4751,726 +4501,1163 @@ mixin RelayoutWhenSystemFontsChangeMixin on RenderObject {
+        if (semantics.parentDataDirty) {
+          continue;
+        }
+        if (!semantics.shouldFormSemanticsNode) {
+          // This render object semantics will need to be merged into a sibling
+          // node.
+          semantics.markNeedsBuild();
         }
       }
-      final List<SemanticsNode> childSiblingNodes = <SemanticsNode>[];
-      fragment.compileChildren(
-        parentSemanticsClipRect: node.parentSemanticsClipRect,
-        parentPaintClipRect: node.parentPaintClipRect,
+    }
+  }
+
+  /// Ensures the semantics nodes from this render object semantics subtree are
+  /// generated and up to date.
+  ///
+  /// This can only be called if the owning rendering object is a semantics
+  /// boundary. For non boundary rendering objects, they require semantics
+  /// information from both their parent and child rendering objects to update
+  /// its cache, so it can't update by themselves.
+  void ensureSemanticsNode() {
+    assert(configProvider.effective.isSemanticBoundary || isRoot);
+    if (isRoot) {
+      buildSemantics(usedSemanticsIds: <int>{}, geometry: _SemanticsGeometry.root);
+    } else {
+      assert(built);
+      // parent data and parent geometry don't change, there isn't anything to
+      // update for semantics nodes generated in this render object semantics.
+      buildSemanticsSubtree(
+        geometry: _SemanticsGeometry(
+          transform: Matrix4.identity(),
+          paintClipRect: cachedSemanticsNode!.parentPaintClipRect,
+          semanticsClipRect: cachedSemanticsNode!.parentSemanticsClipRect,
+        ),
+        usedSemanticsIds: <int>{},
         elevationAdjustment: 0.0,
-        result: children,
-        siblingNodes: childSiblingNodes,
       );
-      siblingNodes.addAll(childSiblingNodes);
     }
+  }
 
-    if (_config.isSemanticBoundary) {
-      owner.assembleSemanticsNode(node, _config, children);
-    } else {
-      node.updateWith(config: _config, childrenInInversePaintOrder: children);
-    }
-    result.add(node);
-    // Sibling node needs to attach to the parent of an explicit node.
-    for (final SemanticsNode siblingNode in siblingNodes) {
-      // sibling nodes are in the same coordinate of the immediate explicit node.
-      // They need to share the same transform if they are going to attach to the
-      // parent of the immediate explicit node.
-      assert(siblingNode.transform == null);
-      siblingNode.transform = node.transform;
-      if (_tagsForChildren != null) {
-        siblingNode.tags ??= <SemanticsTag>{};
-        siblingNode.tags!.addAll(_tagsForChildren!);
+  void didUpdateParentData(_SemanticsParentData newParentData) {
+    if (parentData == newParentData) {
+      return;
+    }
+    // Parent data changes may result in node formation changes.
+    markNeedsBuild();
+    parentData = newParentData;
+    updateChildren();
+  }
+
+  void updateChildren() {
```

Comment:
Maybe mark in this method what's part of phase 1 and phase 2 (from the doc comment)?

This method itself would probably also benefit from a doc comment.